### PR TITLE
Generic JWT feature

### PR DESCRIPTION
**Pull request by yivo**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -70,6 +70,9 @@ gem 'unread', github: 'peatio/unread'
 gem 'carrierwave', '~> 0.10.0'
 gem 'simple_captcha2', require: 'simple_captcha'
 gem 'rest-client', '~> 1.6.8'
+gem 'memoist', '~> 0.16'
+gem 'jwt', '~> 1.2.0'
+gem 'faraday', '~> 0.12.0'
```
Comment (mod) [inline Gemfile:75]: not yet we will do a jwt branch

File: app/api/api_v2/auth/authenticator.rb [hunk 2]
```diff
@@ -1,71 +1,100 @@
...
         @request = request
         @params  = params
       end
 
       def authenticate!
-        check_token!
-        check_tonce!
-        check_signature!
-        token
+        if @request.env['HTTP_AUTHORIZATION'].present?
```
Comment (mod) [inline app/api/api_v2/auth/authenticator.rb:10]: agreed but I think we should have a cleaner JWT check methods
and 
```
begin 
try jwt
try sdk
end
```

File: app/api/api_v2/auth/authenticator.rb [hunk 3]
```diff
@@ -1,71 +1,100 @@
...
+          check_token!
+          check_tonce!
+          check_signature!
+          self.token
+        end
       end
 
       def token
-        @token ||= APIToken.joins(:member).where(access_key: @params[:access_key]).first
+        @token ||= APIToken.joins(:member).where(access_key: access_key).first
```
Comment (mod) [inline app/api/api_v2/auth/authenticator.rb:29]: ok

File: app/api/api_v2/auth/authenticator.rb [hunk 4]
```diff
@@ -1,71 +1,100 @@
...
-        Utils.cache.write key, tonce, 61 # forget after 61 seconds
 
-        now = Time.now.to_i*1000
-        if tonce < now-30000 || tonce > now+30000 # within 30 seconds
+        # Mark tonce as used and remove it from cache after 61 seconds.
+        Utils.cache.write(key, tonce, 61)
+
+        now = Time.now.to_f * 1000
+        # Ensure tonce's value is within 30 seconds.
+        if tonce < now - 30000 || tonce > now + 30000
```
Comment (mod) [inline app/api/api_v2/auth/authenticator.rb:71]: ok

File: app/api/api_v2/auth/authenticator.rb [hunk 5]
```diff
@@ -1,71 +1,100 @@
...
         end
       end
 
-      def tonce
-        @tonce ||= @params[:tonce].to_i
-      end
-
       def payload
-        "#{canonical_verb}|#{APIv2::Mount::PREFIX}#{canonical_uri}|#{canonical_query}"
+        "#{canonical_verb}|#{APIv2::Mount::PREFIX}#{canonical_path}|#{canonical_params}"
```
Comment (mod) [inline app/api/api_v2/auth/authenticator.rb:78]: maybe

File: app/api/api_v2/auth/authenticator.rb [hunk 6]
```diff
@@ -1,71 +1,100 @@
...
 
-      def canonical_uri
+      def canonical_path
         @request.path_info
       end
 
-      def canonical_query
-        hash = @params.select {|k,v| !%w(route_info signature format).include?(k) }
-        URI.unescape(hash.to_param)
+      def canonical_params
```
Comment (mod) [inline app/api/api_v2/auth/authenticator.rb:89]: it seems a big change and it should work without those change
what is it for ?

File: app/api/api_v2/auth/middleware.rb [hunk 7]
```diff
@@ -10,7 +9,9 @@ def before
       end
 
       def provided?
-        params[:access_key] && params[:tonce] && params[:signature]
+        env['HTTP_AUTHORIZATION'].present? ||
```
Comment (mod) [inline app/api/api_v2/auth/middleware.rb:12]: for the next patch.
I personally don't use HTTP_AUTHORIZATION for JWT

File: app/api/api_v2/auth/utils.rb [hunk 8]
```diff
@@ -17,9 +18,8 @@ def urlsafe_string_40
         alias :generate_secret_key :urlsafe_string_40
 
         def hmac_signature(secret_key, payload)
-          OpenSSL::HMAC.hexdigest 'SHA256', secret_key, payload
+          OpenSSL::HMAC.hexdigest('SHA256', secret_key, payload)
         end
```
Comment (mod) [inline app/api/api_v2/auth/utils.rb:22]: ok

File: app/api/api_v2/constraints.rb [hunk 9]
```diff
@@ -16,7 +15,6 @@ def apply_rules!
           req.env['api_v2.token'] && req.env['api_v2.token'].access_key
         end
       end
-
     end
```
Comment (mod) [inline app/api/api_v2/constraints.rb:18]: ok

File: app/api/api_v2/deposits.rb [hunk 10]
```diff
@@ -8,12 +8,11 @@ class Deposits < Grape::API
 
     desc 'Get your deposits history.'
     params do
-      use :auth
       optional :currency, type: String, values: Currency.all.map(&:code), desc: "Currency value contains  #{Currency.all.map(&:code).join(',')}"
       optional :limit, type: Integer, range: 1..100, default: 3, desc: "Set result limit."
       optional :state, type: String, values: Deposit::STATES.map(&:to_s)
     end
-    get "/deposits" do
+    get '/deposits' do
```
Comment (mod) [inline app/api/api_v2/deposits.rb:15]: ok

File: app/api/api_v2/members.rb [hunk 11]
```diff
@@ -2,14 +2,11 @@ module APIv2
   class Members < Grape::API
     helpers ::APIv2::NamedParams
 
+    before { authenticate! }
```
Comment (mod) [inline app/api/api_v2/members.rb:5]: interesting maybe now maybe later you pick you try

File: app/api/api_v2/named_params.rb [hunk 12]
```diff
@@ -2,12 +2,6 @@ module APIv2
   module NamedParams
     extend ::Grape::API::Helpers
 
-    params :auth do
-      requires :access_key, type: String,  desc: "Access key."
-      requires :tonce,      type: Integer, desc: "Tonce is an integer represents the milliseconds elapsed since Unix epoch."
-      requires :signature,  type: String,  desc: "The signature of your request payload, generated using your secret key."
-    end
```
Comment (mod) [inline app/api/api_v2/named_params.rb:9]: I see, we can delay this change in JWT patch

File: app/api/api_v2/orders.rb [hunk 13]
```diff
@@ -35,7 +35,7 @@ class Orders < Grape::API
 
     desc 'Create multiple sell/buy orders.', scopes: %w(trade)
     params do
-      use :auth, :market
+      use :market
```
Comment (mod) [inline app/api/api_v2/orders.rb:38]: I don't understand this change

File: app/api/api_v2/websocket_protocol.rb [hunk 14]
```diff
@@ -1,3 +1,5 @@
+require 'securerandom'
```
Comment (mod) [inline app/api/api_v2/websocket_protocol.rb:1]: try removing the require

File: app/exceptions/auth0/bad_token_error.rb [hunk 15]
```diff
@@ -0,0 +1,5 @@
+module Auth0
```
Comment (mod) [inline app/exceptions/auth0/bad_token_error.rb:1]: not yet

File: app/models/token.rb [hunk 16]
```diff
@@ -51,6 +51,6 @@ def check_latest_send
 
   def generate_token
     self.token = SecureRandom.hex(16)
-    self.expire_at = 30.minutes.from_now
+    self.expires_at = 30.minutes.from_now
```
Comment (mod) [inline app/models/token.rb:54]: ok

File: app/services/auth0/jwt.rb [hunk 17]
```diff
@@ -0,0 +1,31 @@
+require 'openssl/x509'
```
Comment (mod) [inline app/services/auth0/jwt.rb:1]: no yet; and not later we need to fetch using the oauth token through another patch

File: app/services/auth0/peatio_api_token_adapter.rb [hunk 18]
```diff
@@ -0,0 +1,29 @@
+module Auth0
```
Comment (mod) [inline app/services/auth0/peatio_api_token_adapter.rb:1]: so much code here, I am not sure we would need it

**Review by mod — CHANGES_REQUESTED** [review]

**Comment by yivo** [comment]
@mod  Closing this as it contains too much rejected stuff

**Pull request closed without merging**